File: leetcode/Insert_Interval.java
import java.util.*;

/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */

public class Solution {
    boolean intersect(Interval a, Interval b) {
        if (b.start >= a.start && b.start <= a.end) return true;
        if (b.end >= a.start && b.end <= a.end) return true;
        if (a.start >= b.start && a.start <= b.end) return true;
        return false;
    }
    void merge(Interval a, Interval b) {
        b.start = Math.min(a.start, b.start);
        b.end = Math.max(a.end, b.end);
    }
    public ArrayList<Interval> insert(ArrayList<Interval> intervals, Interval newInterval) {
        ArrayList<Interval> ret = new ArrayList<Interval>();
        Object[] inter = intervals.toArray();
        int i = 0;
        while (i < inter.length && ((Interval)inter[i]).end < newInterval.start) {
            ret.add((Interval)inter[i]);
            ++i;
        }
        while (i < inter.length && intersect((Interval)inter[i], newInterval)) {
            merge((Interval)inter[i], newInterval);
            ++i;
        }
        ret.add(newInterval);
        while (i < inter.length && ((Interval)inter[i]).start > newInterval.end) {
            ret.add((Interval)inter[i]);
            ++i;
        }
        return ret;
    }
}

File: leetcode/Insert_Interval.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> insert(vector<Interval> &intervals, Interval newInterval) {
        vector<Interval> after_insertion;
        bool has_inserted = false;  // Need this flag
        for (int i = 0; i < intervals.size(); ++i)
            if (intervals[i].end < newInterval.start) {
                after_insertion.push_back(intervals[i]);
            } else if (intervals[i].start > newInterval.end) {
                if (!has_inserted) {
                    has_inserted = true;
                    after_insertion.push_back(newInterval);
                }
                after_insertion.push_back(intervals[i]);
            } else {
                newInterval.start = min(newInterval.start, intervals[i].start);
                newInterval.end   = max(newInterval.end,   intervals[i].end);
            }
        if (!has_inserted)
            after_insertion.push_back(newInterval);
        return after_insertion;
    }
};
